### PR TITLE
Enum Support - Java Code Generation

### PR DESCRIPTION
**Pull request by TanayParikh**
- Added Enum code generation support for Java
- Updated src/foam/java/Enum.js starter file
- Added refinements in src/foam/java/refinements.js
- Updated  src/foam/java/Class.js and src/foam/java/Method.js to facilitate enum definitions

### CODE REVIEW

**Comment by kgrgreer** [comment]
@adamvy Can you please review. Thanks

**Comment by adamvy** [comment]
@kgrgreer Only if you review #318 :)

File: src/foam/java/Enum.js [hunk 1]
```diff
@@ -1,73 +1,55 @@
...
     else {
-%>  <%= value.name %>(<%= value.index %>, <%= value.javaValue %>, <%= labelForValue(value) %>) <%
+%>  <%= value.name %>(<%= i %>, <%= value.definition.ordinal %>, <%= labelForValue(value) %>) <%
     }
     if ( i == this.values.length - 1 ) {%>;<%}
     else {%>,<%}
   }
 %>
 
   private final int index_;
```
Comment (adamvy) [inline src/foam/java/Enum.js]: Do we really need index_ and ordinal_ ?  Ordinal should be sufficient.

File: src/foam/java/Enum.js [hunk 2]
```diff
@@ -1,73 +1,55 @@
-// Test with: foam.u2.ControllerMode.model_.javaSource();
+// Defines template for the EnumModel
```
Comment (adamvy) [inline src/foam/java/Enum.js]: Overall I wonder if we can do some better code sharing by making a foam.java.Enum class, and having it extend foam.java.Class

An Enum is in essence a Class with a fixed number of implementations.  If we extend foam.java.Class and update the outputJava method to output the VALUES, as well as whatever tweaks we need to match java enum syntax, we should also get methods/fields for free.


I would start by defining foam.java.Enum as a foam.CLASS() that extends foam.java.Class, and then trying to get the output correct.  Then refine buildJavaClass() onto foam.core.EnumModel, which should build a foam.java.Enum.

**Review by adamvy — CHANGES_REQUESTED** [review]

File: src/foam/java/Enum.js (hunk 1 above)
Comment (TanayParikh) [reply src/foam/java/Enum.js]: Thanks, that's a good point. I've removed the index property.

File: src/foam/java/Enum.js [hunk 3]
```diff
@@ -1,97 +1,64 @@
...
-      `
-    }
-  ]
-})
-
-/*
-package <%= this.package %>;
 <%
 var labelForValue = function(value) {
   if (this.androidResource) {
```
Comment (kgrgreer) [inline src/foam/java/Enum.js]: Do we need this now? When we do, can we integrate it with FOAM's Message axiom support and then integrate the message support with Android Resources.

Comment (TanayParikh) [reply src/foam/java/Enum.js]: The androidResource is directly from the FOAM 1 enum template. Should I just go ahead and remove it?

Comment (kgrgreer) [reply src/foam/java/Enum.js]: Yes, just remove for now.

Comment (TanayParikh) [reply src/foam/java/Enum.js]: Removed

File: src/foam/java/Enum.js (hunk 2 above)
Comment (TanayParikh) [reply src/foam/java/Enum.js]: Hi Adam,

I've made foam.java.Enum and extended it from foam.java.Class. I've also updated the general code generation logic to move away from templates and more along the lines of the FOAM2 structure.

File: src/foam/java/Enum.js [hunk 4]
```diff
@@ -1,106 +1,166 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 
 foam.CLASS({
-  refines: 'foam.core.EnumModel',
+  package: 'foam.java',
+  name: 'EnumModel',
```
Comment (adamvy) [inline src/foam/java/Enum.js]: This can just be called Enum, same way foam.java.Class is call Class not ClassModel.

File: src/foam/java/Enum.js [hunk 5]
```diff
@@ -1,106 +1,166 @@
...
+  extends: 'foam.java.Class',
 
-  methods: [
+  properties: [
+    ['javaType', 'java.lang.Enum'],
```
Comment (adamvy) [inline src/foam/java/Enum.js]: Are these used anywhere?  I'm pretty sure these properties should be refined on foam.core.Enum

File: src/foam/java/Enum.js [hunk 6]
```diff
@@ -1,106 +1,166 @@
...
+      o.indent();
+      o.out('}\n');
+
+      this.extras.forEach(function(c) { o.out(c, '\n'); });
+      o.decreaseIndent();
+      o.indent();
+      o.out('}');
+    },
+
+    function toJavaSource() {
```
Comment (adamvy) [inline src/foam/java/Enum.js]: This should be inherited from foam.java.Class

File: src/foam/java/Enum.js (hunk 5 above)
Comment (adamvy) [reply src/foam/java/Enum.js]: Ah reading further, you did refine them onto foam.core.Enum, which is good, these properties here should probably be deleted.

File: src/foam/java/refinements.js [hunk 7]
```diff
@@ -483,13 +484,66 @@ foam.CLASS({
...
+    ['javaType', 'java.lang.Enum'],
+    ['javaInfoType', 'foam.core.AbstractFObjectPropertyInfo'],
+    ['javaJSONParser', 'foam.lib.json.FObjectParser']
+  ]
+});
+
+foam.CLASS({
+  refines: 'foam.core.AbstractEnum',
+
+  properties: [
```
Comment (adamvy) [inline src/foam/java/refinements.js]: I don't believe these properties are needed here.

File: src/foam/java/refinements.js [hunk 8]
```diff
@@ -483,13 +484,66 @@ foam.CLASS({
...
+        };
+      }
+    }
+  ]
+});
+
+foam.CLASS({
+  refines: 'foam.core.EnumModel',
+
+  properties: [
```
Comment (adamvy) [inline src/foam/java/refinements.js]: Again with these properties :), I don't think you need them here either

File: src/foam/java/Enum.js [hunk 9]
```diff
@@ -1,106 +1,166 @@
...
+  extends: 'foam.java.Class',
 
-  methods: [
+  properties: [
+    ['javaType', 'java.lang.Enum'],
+    ['javaInfoType', 'foam.core.AbstractFObjectPropertyInfo'],
+    ['javaJSONParser', 'foam.lib.json.FObjectParser'],
   ],
+  methods: [
+    function outputJava(o) {
```
Comment (adamvy) [inline src/foam/java/Enum.js]: This is a good start.

Next I would look at if you can refactor this to share some code with outputJava in foam.java.Class.  There is a fair amount of duplication here that can probably be cleaned up.

**Comment by kirkeaton** [comment]
@adamvy Are there any more changes required for this PR?

**Comment by adamvy** [comment]
Sorry for the delay, this looks good to me, thanks!

**Pull request merged**